dff/utils/benchmark/context_storage.py

```diff
@@ -0,0 +1,585 @@
+            "result" field should be a dictionary with the values returned by
+            :py:func:`~.time_context_read_write` and keys
+            "write_times", "read_times" and "update_times".
+        :return: None
+        """
+        if not benchmark["success"] or isinstance(benchmark["result"], str):
+            return
+
+        def get_complex_stats(results):
+            if len(results) == 0 or len(results[0]) == 0:
+                return [], {}, None
+
+            average_grouped_by_context_num = [mean(times.values()) for times in results]
+            average_grouped_by_dialog_len = {key: mean([times[key] for times in results]) for key in results[0].keys()}
+            average = float(mean(average_grouped_by_context_num))
+            return average_grouped_by_context_num, average_grouped_by_dialog_len, average
+
+        read_stats = get_complex_stats(benchmark["result"]["read_times"])
+        update_stats = get_complex_stats(benchmark["result"]["update_times"])
+
+        result = {
+            "average_write_time": mean(benchmark["result"]["write_times"]),
+            "average_read_time": read_stats[2],
+            "average_update_time": update_stats[2],
+            "read_times_grouped_by_context_num": read_stats[0],
+            "read_times_grouped_by_dialog_len": read_stats[1],
+            "update_times_grouped_by_context_num": update_stats[0],
+            "update_times_grouped_by_dialog_len": update_stats[1],
+        }
+        result["pretty_write"] = (
+            float(f'{result["average_write_time"]:.3}') if result["average_write_time"] is not None else None
+        )
+        result["pretty_read"] = (
+            float(f'{result["average_read_time"]:.3}') if result["average_read_time"] is not None else None
+        )
+        result["pretty_update"] = (
+            float(f'{result["average_update_time"]:.3}') if result["average_update_time"] is not None else None
+        )
+        result["pretty_read+update"] = (
+            float(f'{result["average_read_time"] + result["average_update_time"]:.3}')
+            if result["average_read_time"] is not None and result["average_update_time"] is not None
+            else None
+        )
+
+        benchmark["average_results"] = result
+
+    def _run(self):
+        try:
+            write_times, read_times, update_times = time_context_read_write(
+                self.db_factory.db(),
+                self.benchmark_config.get_context(),
+                self.benchmark_config.context_num,
+                context_updater=self.benchmark_config.get_context_updater(),
+            )
+            return {
+                "success": True,
+                "result": {
+                    "write_times": write_times,
+                    "read_times": read_times,
+                    "update_times": update_times,
+                },
+            }
+        except Exception as e:
+            exception_message = getattr(e, "message", repr(e))
```

Comment:
What is actually the point? Something that is not an exception with a message can be thrown **and** catched here, like in C?